### DISCUSSION TITLE
Open a file that was just created through a shell command

### DISCUSSION THREAD

**Discussion by ReaderGuy42** [Taitava/obsidian-shellcommands#395] [Q&A] [answered]
I'm having trouble figuring out which reply goes with which of my many questions in this discussion here https://github.com/Taitava/obsidian-shellcommands/discussions/369#discussioncomment-7807357 so I wanted to ask this again.

First, thanks so much for creating this plugin, it's super helpful for me, and thanks also for taking the time to reply!

I have a shell command: 
`cd {{folder_path:absolute}}; bash "Scripts/date-iterate.sh" {{file_name}}`

This runs the script `date-iterate.sh` on the target Obsidian note.
```
#!/bin/bash

fileName=$1

today=$(date +"%Y-%m-%d")


if [[ $fileName =~ ([0-9]+-[0-9]+-[0-9]+)(.*)([0-9]+)\.([^.]*) ]]; then
  # BASH_REMATCH match groups:
  # 1 - date
  # 2 - everything else
  # 3 - number
  # 4 - extension

  number=$(( ${BASH_REMATCH[3]} + 1 ))

  while :; do

        newFileName="$(date +%Y-%m-%d)${BASH_REMATCH[2]}${number}.${BASH_REMATCH[4]}"

        [[ -f $newFileName ]] || break

        let number++

  done

  echo "copy $fileName to $newFileName"

  cp "$fileName" "$newFileName"

elif [[ $fileName =~ (.*)\.([^.]*) ]]; then

    newNumber="1"

      addInfoName="$today ${BASH_REMATCH[1]} V.${newNumber}.${BASH_REMATCH[2]}"

          mv "$fileName" "$addInfoName" 

                let newNumber++

           cp "$addInfoName" "$today ${BASH_REMATCH[1]} V.${newNumber}.${BASH_REMATCH[2]}"
                 
fi
```

I don't remember exactly who helped me with this script, it might have been you or someone on another forum, either way it works as intended.

It takes files and renames them.
`filename` gets turned into `2024-02-16 filename V.1` and then also creates a copy `2024-02-16 filename V.2`. If there's already a date at the start of the file it just creates a copy and changes the date to the current one, e.g.  `2024-02-10 filename V.1` gets copied and the new file then is  `2024-02-16 filename V.2`. This is my own way of version control within Obsidian.

However, to add the cherry on top, I'd like it to automatically navigate to that new file. 

I've tried following your solution here https://github.com/Taitava/obsidian-shellcommands/discussions/369#discussioncomment-7036730 but when I set my `stdout` to Open Files and echo the file path, and I add `echo {{file_path:absolute}}` to my shell command, so then the whole command looks like this: 
`cd {{folder_path:absolute}}; bash "Scripts/date-iterate.sh" {{file_name}};
echo {{file_path:absolute}}`

But then I get this error pop up:
![Screenshot from 2024-02-16 12-35-59](https://github.com/Taitava/obsidian-shellcommands/assets/71895400/531b4f6d-c641-4d32-8a83-9cef28be2fc2)

It’s somehow taking the command to copy as a file name I guess, though I don't know where it's getting that from.

Any advice is appreciated. Thanks!!

**Comment by Taitava** [comment, marked as answer]
Hi! 🙂 Thank you for your question and nice feedback! 🌞 

The problem is that `date-iterate.sh` is outputting the text `copy 2024-02-10 example V.2.md to 2024-02-16 example V.6.md` to `stdout`:
```bash
echo "copy $fileName to $newFileName"
```

To fix this, you could change the `echo` line to only include the new file name:
```bash
echo "$newFileName"
```

Then, if you still would like to see a notification balloon in Obsidian showing the text that was originally echoed, you can output that text to `stderr`:
```bash
echo "$newFileName"
echo "copy $fileName to $newFileName" 1>&2
```

So, this is what happens now:
- The [Open files](https://publish.obsidian.md/shellcommands/Output+handling/Output+channel+-+Open+files) output handler reads the text outputted to `stdout` (standard output) by the `echo` line that **does not** contain `1>&2`. _Open files_ does not read `stderr` (standard error) so we can echo anything to `stderr` without breaking _Open files_.
- On the other `echo` line, [`1>&2` directs the outputted text to `stderr`](https://tldp.org/HOWTO/Bash-Prog-Intro-HOWTO-3.html#ss3.4) (instead of `stdout`), which the SC plugin displays in a notification balloon (by default, if you haven't [changed the output channel](https://publish.obsidian.md/shellcommands/Output+handling/Output+channels#How+to+select+output+channels+for+a+shell+command) for `stderr`).

Please let me know if this helped you, or if you still have any questions. 👍

---

Here are the above-mentioned changes edited in your `date-iterate.sh` script:
```bash
#!/bin/bash

fileName=$1

today=$(date +"%Y-%m-%d")


if [[ $fileName =~ ([0-9]+-[0-9]+-[0-9]+)(.*)([0-9]+)\.([^.]*) ]]; then
  # BASH_REMATCH match groups:
  # 1 - date
  # 2 - everything else
  # 3 - number
  # 4 - extension

  number=$(( ${BASH_REMATCH[3]} + 1 ))

  while :; do

        newFileName="$(date +%Y-%m-%d)${BASH_REMATCH[2]}${number}.${BASH_REMATCH[4]}"

        [[ -f $newFileName ]] || break

        let number++

  done

  # Open the new file in Obsidian:
  echo "$newFileName"

  # Show a notification baloon in Obsidian:
  echo "copy $fileName to $newFileName" 1>&2

  cp "$fileName" "$newFileName"

elif [[ $fileName =~ (.*)\.([^.]*) ]]; then

    newNumber="1"

      addInfoName="$today ${BASH_REMATCH[1]} V.${newNumber}.${BASH_REMATCH[2]}"

          mv "$fileName" "$addInfoName" 

                let newNumber++

           cp "$addInfoName" "$today ${BASH_REMATCH[1]} V.${newNumber}.${BASH_REMATCH[2]}"

           # Open the new file in Obsidian:
           echo "$today ${BASH_REMATCH[1]} V.${newNumber}.${BASH_REMATCH[2]}"
                 
fi
```

Edit: I also added the following to the `elif` block:
```bash
# Open the new file in Obsidian:
echo "$today ${BASH_REMATCH[1]} V.${newNumber}.${BASH_REMATCH[2]}"
```

**Reply by Taitava** [reply]
## Also, one note about this:

> when I set my `stdout` to Open Files and echo the file path, and I add `echo {{file_path:absolute}}` to my shell command, so then the whole command looks like this: `cd {{folder_path:absolute}}; bash "Scripts/date-iterate.sh" {{file_name}}; echo {{file_path:absolute}}`

Doing `echo {{file_path:absolute}}` would not work here, because [{{file_path:absolute}}](https://publish.obsidian.md/shellcommands/Variables/%7B%7Bevent_file_path%7D%7D) would give the original (old) file name, not the new name. This is because **all** variables (like `{{file_path:absolute}}`) are resolved _before_ the shell command is executed, i.e. before the copying.

`echo {{file_path:absolute}}` would reopen the original file.

**Reply by ReaderGuy42** [reply]
Thank you so much! All I had to do was change the change the `echo` command in my shell script to `echo "$newFileName"`. I completely forgot to look through my script to find where that `copy` error was coming from. Now it works!! Thanks :)

**Reply by ReaderGuy42** [reply]
Hi again, sorry to be a bother, but it's somehow working sometimes but not always, which is confusing me. I'm now sometimes getting this error:
![image](https://github.com/Taitava/obsidian-shellcommands/assets/71895400/d4fb8e82-a125-4b87-b32c-e00e5093ccaa)

The error is strange, since the file is created correctly, it just doesn't see it?

**Reply by Taitava** [reply]
Hi, does it help if you:
- Add `sleep 1` after the `cp` command (either in your script file or in the shell command after calling the script file). It will delay the shell command's completion by one second. Therefore it will delay the attempt to open the file.
- When doing this, please make sure the shell command's [output handling mode is _Wait until finished_, not _Realtime_](https://publish.obsidian.md/shellcommands/Output+handling/Realtime+output+handling#Wait+until+finished).

This can help **if** the file creation does not get complete before the shell command's execution finishes, i.e. if the file doesn't have enough time to be born before the SC plugin tries to read it. I'm not sure if this is the case for you, but it's one thing than can be tried. If it doesn't help, we can continue searching for a solution.

P.S. I know `sleep 1` can be a bit annoying as it slows down the opening. You can try e.g. `sleep 0.5` to make it a bit quicker.

**Reply by ReaderGuy42** [reply]
Yup, with `sleep 0.5` it's perfect, thank you!!
I got some help optimizing the script over here: https://stackoverflow.com/a/78031222/15999132
